sklearn_numba_dpex/common/kernels.py

```diff
@@ -216,46 +216,70 @@ def make_sum_reduction_2d_kernel(
-    axis=0. This is repeated as many times as needed until only one value remains in
-    global memory.
-
-    If `fused_elementwise_func` is not None, it will be applied element-wise before
-    summing. It must be a function that will be interpreted as a dpex.func and is
-    subject to the same rules ( see
+    numba_dpex does not provide tools such as `cuda.reduce` so we implement
+    from scratch a reduction strategy. The strategy relies on the associativoty
+    of the operation used for the reduction, thus allowing to reduce the input
+    in any order.
+
+    The strategy consists in performing a series of kernel invocations that
+    each perform a partial sum. At each kernel invocation, the input array is
+    tiled with non-overlapping windows and all the values within a given window
+    are collaboratively summed by the work items of a given work group.
+
+    Each window has shape:
+
+    - when `axis = 0`: `(reduction_block_size, 1)` with: `reduction_block_size
+      = work_group_size // sub_group_size * 2`
+
+    - when `axis = 1`: `(1, reduction_block_size)` with: `reduction_block_size
+      = work_group_size * 2`
+
+    Once the reduction is done in a work group the result is written back to
+    global memory, thus creating an intermediary result array whose size is
+    divided by `reduction_block_size`.
+
+    This is repeated as many times as needed until only one value remains per
+    column if axis=0 or per row if axis=1.
+
+    During a single kernel invocation, several iterations of local reductions
+    are performed using local memory to store intermediate results. At each
+    local iteration, each work item combines two values, thus halving the
+    number of values, and the number of active work items for the subsequent
+    iterations. Work items are progressively discarded in a bracket manner. The
+    work items with the greatest ids are discarded first, and we rely on the
+    fact that the remaining work items are adjacents to optimize the read-write
+    operations in local memory.
+
+    If `fused_elementwise_func` is not None, it will be applied element-wise
+    once to each element of the input array at the beginning of the the first
+    kernel invocation. This function is compiled and fused into the first
+    kernel as a device function with the help of `dpex.func`. This comes with
+    limitations as explained in:
+
     https://intelpython.github.io/numba-dpex/latest/user_guides/kernel_programming_guide/device-functions.html # noqa
-    ). It is expected to take one scalar argument and returning one scalar value.
-    lambda functions are advised against since the cache will not work with lambda
-    functions. sklearn_numba_dpex.common._utils expose some pre-defined
-    `fused_elementwise_func`s.
+
+    It is expected to take one scalar argument and returning one
+    scalar value. lambda functions are advised against since there compilation
+    cannot be cached.
```

Comment:
In fact the cache will work as long as the hash taken by `lru_cache` from the signature is the same. Two different lambda functions, even if they define the same function, will have different hashes (probably hashes with `id(func)` ?).  So the cache will work if a given lambda function is defined once and re-used, and stop working as soon as it's re-defined. :shrug: not sure it's worth mentioning since I'd imagine cache failing in this case can be expected.

So maybe be more vague:

```suggestion
    might not be cached.
```